app/assets/javascripts/discourse/app/components/emoji-picker.js

```diff
@@ -245,15 +255,152 @@ export default Component.extend({
+    const firstSection = document.querySelectorAll(
+      this.elements.firstEmojiSection
+    );
+
+    let currentEmoji;
+
+    this.set(
+      "hoveredEmoji",
+      this._codeWithDiversity(event.target.title, this.selectedDiversity)
+    );
+
+    // if no search results get all emojis
+    if (emojis.length === 0) {
+      emojis = document.querySelectorAll(this.elements.allEmojis);
+    }
+
+    if (
+      event.code === "ArrowDown" &&
+      this._focusedOn(this.elements.searchBar)
+    ) {
+      return emojis[0].focus();
+    }
+
     if (event.code === "Escape") {
       this.onClose(event);
       return false;
     }
+
+    if (arrowKeys.includes(event.code)) {
+      if (!this._focusedOn(this.elements.picker)) {
+        return;
+      }
+
+      Array.from(emojis).find((e, index) => {
+        currentEmoji = index;
+        return e.isEqualNode(event.target);
+      });
+
+      const numEmojisInRow = this.get("emojiPerRow");
+
+      if (event.code === "ArrowRight") {
+        let nextEmoji = currentEmoji + 1;
+
+        if (nextEmoji < emojis.length) {
+          emojis[nextEmoji].focus();
+        } else if (nextEmoji >= emojis.length) {
+          nextEmoji = 0;
+          emojis[nextEmoji].focus();
+        }
+      }
+
+      if (event.code === "ArrowLeft") {
+        const previousEmoji = currentEmoji - 1;
+        if (currentEmoji > 0) {
+          emojis[previousEmoji].focus();
+        }
+      }
+
+      if (event.code === "ArrowDown") {
+        const emojiNextRow = currentEmoji + numEmojisInRow;
+
+        if (
+          event.target.classList.contains("recent-emoji") &&
+          recentEmojis.length < 11
+        ) {
+          firstSection[0].focus();
+        } else {
+          if (emojiNextRow < emojis.length) {
+            emojis[emojiNextRow].focus();
+          }
+        }
+      }
+
+      if (event.code === "ArrowUp") {
+        const emojiPreviousRow = currentEmoji - numEmojisInRow;
+        const emojiInFirstRow = 0 + numEmojisInRow;
+
+        if (emojiPreviousRow >= 0) {
+          emojis[emojiPreviousRow].focus();
+        }
+
+        if (currentEmoji < emojiInFirstRow) {
+          searchBar.focus();
+        }
+      }
+
+      event.preventDefault();
+      return false;
+    }
+
+    if (event.code === "Enter") {
+      if (!this._focusedOn(".emoji")) {
+        return;
+      }
+      this.onEmojiSelection(event);
+      this.onClose(event);
+      event.preventDefault();
+      return false;
+    }
   },
 
   @action
   onFilterChange(event) {
     this._applyFilter(event.target.value);
+
+    discourseDebounce(
+      this,
+      () => {
+        if (event.target.value === "") {
+          this._getNumEmojiPerRow(this.elements.emojiNoRecents, false);
+        } else {
+          this._getNumEmojiPerRow(this.elements.emojiResults, true);
+        }
+      },
+      500
+    );
+  },
+
+  _getNumEmojiPerRow(emojiSelector, isSearching = false) {
```

Comment:
Another naming suggestion, but `_getNumEmojiPerRow` feels wrong here because this is _setting_ the `emojiPerRow` property. So maybe use `_setEmojiPerRow` or `_calcEmojiPerRow`.